### ISSUE TITLE
Not suitable for production use, leaks subscriptions and prevents components being GCd when AOT compiled

### ISSUE THREAD

**Issue by ohjames** [pldin601/useful-angular-decorators#1]
@pldin601 the problem with your code is that it dynamically adds the `ngOnDestroy` lifecycle method. The AOT compiled code only calls `ngOnDestroy` if the compiler could statically infer its existence. So your method will only work if every component that uses the decorator also defines an `ngOnDestory` (it can be empty if not needed). In all other cases it will leak subscriptions, preventing components from being garbage collected.

**Comment by pldin601** [comment]
Thanks! Will add this check on decorator initialization.

**Comment by ohjames** [comment]
Not fixed, putting your subjects in a weak map won't complete them, so the component memory still can't be garbage collected, and your object references will ensure the WeakMap never shrinks. Just use `observable-input` instead.

**Comment by ohjames** [comment]
Also the name `angular-decorators` is a really poor choice. Packages should be organised and named by their *features* rather than their *nature*.

**Comment by ohjames** [comment]
Haha thanks to you I noticed observable-input has the exact same memory leak as your package. Doah! So anybody who's used my library or chooses to use your library will suffer from terrible memory leaks.

The only way around this is to force the user to use a superclass (to ensure the destroy logic runs) or to force the user to define their own destroy hook so the AOT compiler knows to call the overridden version.

**Comment by pldin601** [comment]
@ohjames , I thought that I was crazy when I look at your implementation and see the same :-D
Okay. I change behavior to enforce users to create ngOnDestroy method to use decorator. We have few decorators what will be published here so I think name of this repo is ok. Also it is prefixed.

**Comment by ohjames** [comment]
> We have few decorators what will be published here so I think name of this repo is ok

Nope, code should be organised by feature, not nature. Part of angular's own coding guidelines but also just common sense.

**Comment by pldin601** [comment]
Can you share angular coding guideline with me?

**Comment by ohjames** [comment]
https://angular.io/guide/styleguide <-- grep for "Application structure"

Also see https://marmelab.com/blog/2015/12/17/react-directory-structure.html

**Comment by pldin601** [comment]
@ohjames Trying to understand why WeakMap can prevent GCing of components.

As I see here https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap:

> The WeakMap object is a collection of key/value pairs in which the keys are weakly referenced.

There is no place where references can prevent memory leak in decorator. Only one reference to component - key in WeakMap.
